// @epicgames-ps/lib-pixelstreamingfrontend-ue5.5/dist/cjs/Inputs/InputClassesFactory.js
"use strict";
// Copyright Epic Games, Inc. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActiveKeys = exports.InputClassesFactory = void 0;
const KeyboardController_1 = require("./KeyboardController");
const MouseControllerLocked_1 = require("./MouseControllerLocked");
const MouseControllerHovering_1 = require("./MouseControllerHovering");
const TouchController_1 = require("./TouchController");
const TouchControllerFake_1 = require("./TouchControllerFake");
const GamepadController_1 = require("./GamepadController");
const Config_1 = require("../Config/Config");
const lib_pixelstreamingcommon_ue5_5_1 = require("@epicgames-ps/lib-pixelstreamingcommon-ue5.5");
/**
 * Class for making and setting up input class types
 */
class InputClassesFactory {
    /**
     * @param toStreamerMessagesProvider - Stream message instance
     * @param videoElementProvider - Video Player instance
     * @param coordinateConverter - A coordinateConverter instance
     */
    constructor(toStreamerMessagesProvider, videoElementProvider, coordinateConverter) {
        this.activeKeys = new ActiveKeys();
        this.toStreamerMessagesProvider = toStreamerMessagesProvider;
        this.videoElementProvider = videoElementProvider;
        this.coordinateConverter = coordinateConverter;
    }
    /**
     * Registers browser key events.
     */
    registerKeyBoard(config) {
        lib_pixelstreamingcommon_ue5_5_1.Logger.Info('Register Keyboard Events');
        const keyboardController = new KeyboardController_1.KeyboardController(this.toStreamerMessagesProvider, config, this.activeKeys);
        keyboardController.register();
        return keyboardController;
    }
    /**
     * register mouse events based on a control type
     * @param controlScheme - if the mouse is either hovering or locked
     */
    registerMouse(controlScheme) {
        lib_pixelstreamingcommon_ue5_5_1.Logger.Info('Register Mouse Events');
        let mouseController;
        if (controlScheme == Config_1.ControlSchemeType.HoveringMouse) {
            mouseController = new MouseControllerHovering_1.MouseControllerHovering(this.toStreamerMessagesProvider, this.videoElementProvider, this.coordinateConverter, this.activeKeys);
        }
        else {
            mouseController = new MouseControllerLocked_1.MouseControllerLocked(this.toStreamerMessagesProvider, this.videoElementProvider, this.coordinateConverter, this.activeKeys);
        }
        mouseController.register();
        return mouseController;
    }
    /**
     * register touch events
     * @param fakeMouseTouch - the faked mouse touch event
     */
    registerTouch(fakeMouseTouch) {
        lib_pixelstreamingcommon_ue5_5_1.Logger.Info('Registering Touch');
        let touchController;
        if (fakeMouseTouch) {
            touchController = new TouchControllerFake_1.TouchControllerFake(this.toStreamerMessagesProvider, this.videoElementProvider, this.coordinateConverter);
        }
        else {
            touchController = new TouchController_1.TouchController(this.toStreamerMessagesProvider, this.videoElementProvider, this.coordinateConverter);
        }
        touchController.register();
        return touchController;
    }
    /**
     * registers a gamepad
     */
    registerGamePad() {
        lib_pixelstreamingcommon_ue5_5_1.Logger.Info('Register Game Pad');
        const gamepadController = new GamepadController_1.GamepadController(this.toStreamerMessagesProvider);
        gamepadController.register();
        return gamepadController;
    }
}
exports.InputClassesFactory = InputClassesFactory;
/**
 * A class that keeps track of current active keys
 */
class ActiveKeys {
    constructor() {
        this.activeKeys = [];
        this.activeKeys = [];
    }
    /**
     * Get the current array of active keys
     * @returns - an array of active keys
     */
    getActiveKeys() {
        return this.activeKeys;
    }
}
exports.ActiveKeys = ActiveKeys;
