// @epicgames-ps/lib-pixelstreamingfrontend-ue5.5/dist/cjs/Inputs/MouseController.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MouseController = void 0;
// Copyright Epic Games, Inc. All Rights Reserved.
const MouseButtons_1 = require("./MouseButtons");
/**
 * The base class for mouse controllers. Since there is a bunch of shared behaviour between locked and
 * hover mouse controllers this is where that shared behaviour lives.
 */
class MouseController {
    constructor(streamMessageController, videoPlayer, coordinateConverter, activeKeys) {
        this.streamMessageController = streamMessageController;
        this.coordinateConverter = coordinateConverter;
        this.videoPlayer = videoPlayer;
        this.activeKeys = activeKeys;
        this.onEnterListener = this.onMouseEnter.bind(this);
        this.onLeaveListener = this.onMouseLeave.bind(this);
    }
    register() {
        this.registerMouseEnterAndLeaveEvents();
    }
    unregister() {
        this.unregisterMouseEnterAndLeaveEvents();
    }
    registerMouseEnterAndLeaveEvents() {
        const videoElementParent = this.videoPlayer.getVideoParentElement();
        videoElementParent === null || videoElementParent === void 0 ? void 0 : videoElementParent.addEventListener('mouseenter', this.onEnterListener);
        videoElementParent === null || videoElementParent === void 0 ? void 0 : videoElementParent.addEventListener('mouseleave', this.onLeaveListener);
    }
    unregisterMouseEnterAndLeaveEvents() {
        const videoElementParent = this.videoPlayer.getVideoParentElement();
        videoElementParent === null || videoElementParent === void 0 ? void 0 : videoElementParent.removeEventListener('mouseenter', this.onEnterListener);
        videoElementParent === null || videoElementParent === void 0 ? void 0 : videoElementParent.removeEventListener('mouseleave', this.onLeaveListener);
    }
    onMouseEnter(event) {
        var _a;
        if (!this.videoPlayer.isVideoReady()) {
            return;
        }
        (_a = this.streamMessageController.toStreamerHandlers.get('MouseEnter')) === null || _a === void 0 ? void 0 : _a();
        this.pressMouseButtons(event.buttons, event.x, event.y);
    }
    onMouseLeave(event) {
        var _a;
        if (!this.videoPlayer.isVideoReady()) {
            return;
        }
        (_a = this.streamMessageController.toStreamerHandlers.get('MouseLeave')) === null || _a === void 0 ? void 0 : _a();
        this.releaseMouseButtons(event.buttons, event.x, event.y);
    }
    releaseMouseButtons(buttons, X, Y) {
        const coord = this.coordinateConverter.translateUnsigned(X, Y);
        if (buttons & MouseButtons_1.MouseButtonsMask.primaryButton) {
            this.sendMouseUp(MouseButtons_1.MouseButton.mainButton, coord.x, coord.y);
        }
        if (buttons & MouseButtons_1.MouseButtonsMask.secondaryButton) {
            this.sendMouseUp(MouseButtons_1.MouseButton.secondaryButton, coord.x, coord.y);
        }
        if (buttons & MouseButtons_1.MouseButtonsMask.auxiliaryButton) {
            this.sendMouseUp(MouseButtons_1.MouseButton.auxiliaryButton, coord.x, coord.y);
        }
        if (buttons & MouseButtons_1.MouseButtonsMask.fourthButton) {
            this.sendMouseUp(MouseButtons_1.MouseButton.fourthButton, coord.x, coord.y);
        }
        if (buttons & MouseButtons_1.MouseButtonsMask.fifthButton) {
            this.sendMouseUp(MouseButtons_1.MouseButton.fifthButton, coord.x, coord.y);
        }
    }
    pressMouseButtons(buttons, X, Y) {
        if (!this.videoPlayer.isVideoReady()) {
            return;
        }
        const coord = this.coordinateConverter.translateUnsigned(X, Y);
        if (buttons & MouseButtons_1.MouseButtonsMask.primaryButton) {
            this.sendMouseDown(MouseButtons_1.MouseButton.mainButton, coord.x, coord.y);
        }
        if (buttons & MouseButtons_1.MouseButtonsMask.secondaryButton) {
            this.sendMouseDown(MouseButtons_1.MouseButton.secondaryButton, coord.x, coord.y);
        }
        if (buttons & MouseButtons_1.MouseButtonsMask.auxiliaryButton) {
            this.sendMouseDown(MouseButtons_1.MouseButton.auxiliaryButton, coord.x, coord.y);
        }
        if (buttons & MouseButtons_1.MouseButtonsMask.fourthButton) {
            this.sendMouseDown(MouseButtons_1.MouseButton.fourthButton, coord.x, coord.y);
        }
        if (buttons & MouseButtons_1.MouseButtonsMask.fifthButton) {
            this.sendMouseDown(MouseButtons_1.MouseButton.fifthButton, coord.x, coord.y);
        }
    }
    sendMouseDown(button, X, Y) {
        var _a;
        (_a = this.streamMessageController.toStreamerHandlers.get('MouseDown')) === null || _a === void 0 ? void 0 : _a([button, X, Y]);
    }
    sendMouseUp(button, X, Y) {
        var _a;
        const coord = this.coordinateConverter.translateUnsigned(X, Y);
        (_a = this.streamMessageController.toStreamerHandlers.get('MouseUp')) === null || _a === void 0 ? void 0 : _a([button, coord.x, coord.y]);
    }
}
exports.MouseController = MouseController;
