// @epicgames-ps/lib-pixelstreamingfrontend-ui-ue5.5/dist/cjs/UI/Controls.js
"use strict";
// Copyright Epic Games, Inc. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Controls = void 0;
const FullscreenIcon_1 = require("./FullscreenIcon");
const SettingsIcon_1 = require("./SettingsIcon");
const StatsIcon_1 = require("./StatsIcon");
const XRIcon_1 = require("./XRIcon");
const lib_pixelstreamingfrontend_ue5_5_1 = require("@epicgames-ps/lib-pixelstreamingfrontend-ue5.5");
const UIConfigurationTypes_1 = require("../UI/UIConfigurationTypes");
// If there isn't a type provided, default behaviour is to create the element.
function shouldCreateButton(type) {
    return type == undefined ? true : type.creationMode === UIConfigurationTypes_1.UIElementCreationMode.CreateDefaultElement;
}
/**
 * Element containing various controls like stats, settings, fullscreen.
 */
class Controls {
    /**
     * Construct the controls
     */
    constructor(config) {
        this.config = config;
        if (!config || shouldCreateButton(config.statsButtonType)) {
            this.statsIcon = new StatsIcon_1.StatsIcon();
        }
        if (!config || shouldCreateButton(config.settingsButtonType)) {
            this.settingsIcon = new SettingsIcon_1.SettingsIcon();
        }
        if (!config || shouldCreateButton(config.fullscreenButtonType)) {
            this.fullscreenIcon = new FullscreenIcon_1.FullScreenIcon();
        }
        if (!config || shouldCreateButton(config.xrIconType)) {
            this.xrIcon = new XRIcon_1.XRIcon();
        }
    }
    /**
     * Get the element containing the controls.
     */
    get rootElement() {
        if (!this._rootElement) {
            this._rootElement = document.createElement('div');
            this._rootElement.id = 'controls';
            if (this.fullscreenIcon) {
                this._rootElement.appendChild(this.fullscreenIcon.rootElement);
            }
            if (this.settingsIcon) {
                this._rootElement.appendChild(this.settingsIcon.rootElement);
            }
            if (this.statsIcon) {
                this._rootElement.appendChild(this.statsIcon.rootElement);
            }
            if (this.xrIcon) {
                void lib_pixelstreamingfrontend_ue5_5_1.WebXRController.isSessionSupported('immersive-vr').then((supported) => {
                    if (supported) {
                        this._rootElement.appendChild(this.xrIcon.rootElement);
                    }
                });
            }
            document.addEventListener('fullscreenchange', this.handleFullscreenChange.bind(this));
        }
        return this._rootElement;
    }
    handleFullscreenChange() {
        var _a;
        const isInFullscreen = !!document.fullscreenElement;
        if (isInFullscreen && ((_a = this.config) === null || _a === void 0 ? void 0 : _a.hideControlsInFullscreen)) {
            this._rootElement.style.visibility = 'hidden';
        }
        else {
            this._rootElement.style.visibility = 'visible';
        }
    }
}
exports.Controls = Controls;
