// @epicgames-ps/lib-pixelstreamingfrontend-ue5.5/dist/esm/Inputs/InputClassesFactory.js
// Copyright Epic Games, Inc. All Rights Reserved.
import { KeyboardController } from './KeyboardController';
import { MouseControllerLocked } from './MouseControllerLocked';
import { MouseControllerHovering } from './MouseControllerHovering';
import { TouchController } from './TouchController';
import { TouchControllerFake } from './TouchControllerFake';
import { GamepadController } from './GamepadController';
import { ControlSchemeType } from '../Config/Config';
import { Logger } from '@epicgames-ps/lib-pixelstreamingcommon-ue5.5';
/**
 * Class for making and setting up input class types
 */
export class InputClassesFactory {
    /**
     * @param toStreamerMessagesProvider - Stream message instance
     * @param videoElementProvider - Video Player instance
     * @param coordinateConverter - A coordinateConverter instance
     */
    constructor(toStreamerMessagesProvider, videoElementProvider, coordinateConverter) {
        this.activeKeys = new ActiveKeys();
        this.toStreamerMessagesProvider = toStreamerMessagesProvider;
        this.videoElementProvider = videoElementProvider;
        this.coordinateConverter = coordinateConverter;
    }
    /**
     * Registers browser key events.
     */
    registerKeyBoard(config) {
        Logger.Info('Register Keyboard Events');
        const keyboardController = new KeyboardController(this.toStreamerMessagesProvider, config, this.activeKeys);
        keyboardController.register();
        return keyboardController;
    }
    /**
     * register mouse events based on a control type
     * @param controlScheme - if the mouse is either hovering or locked
     */
    registerMouse(controlScheme) {
        Logger.Info('Register Mouse Events');
        let mouseController;
        if (controlScheme == ControlSchemeType.HoveringMouse) {
            mouseController = new MouseControllerHovering(this.toStreamerMessagesProvider, this.videoElementProvider, this.coordinateConverter, this.activeKeys);
        }
        else {
            mouseController = new MouseControllerLocked(this.toStreamerMessagesProvider, this.videoElementProvider, this.coordinateConverter, this.activeKeys);
        }
        mouseController.register();
        return mouseController;
    }
    /**
     * register touch events
     * @param fakeMouseTouch - the faked mouse touch event
     */
    registerTouch(fakeMouseTouch) {
        Logger.Info('Registering Touch');
        let touchController;
        if (fakeMouseTouch) {
            touchController = new TouchControllerFake(this.toStreamerMessagesProvider, this.videoElementProvider, this.coordinateConverter);
        }
        else {
            touchController = new TouchController(this.toStreamerMessagesProvider, this.videoElementProvider, this.coordinateConverter);
        }
        touchController.register();
        return touchController;
    }
    /**
     * registers a gamepad
     */
    registerGamePad() {
        Logger.Info('Register Game Pad');
        const gamepadController = new GamepadController(this.toStreamerMessagesProvider);
        gamepadController.register();
        return gamepadController;
    }
}
/**
 * A class that keeps track of current active keys
 */
export class ActiveKeys {
    constructor() {
        this.activeKeys = [];
        this.activeKeys = [];
    }
    /**
     * Get the current array of active keys
     * @returns - an array of active keys
     */
    getActiveKeys() {
        return this.activeKeys;
    }
}
