// @epicgames-ps/lib-pixelstreamingfrontend-ue5.5/dist/esm/Config/SettingOption.js
// Copyright Epic Games, Inc. All Rights Reserved.
import { Logger } from '@epicgames-ps/lib-pixelstreamingcommon-ue5.5';
import { SettingBase } from './SettingBase';
/**
 * An Option setting object with a text label. Allows you to specify an array of options and select one of them.
 */
export class SettingOption extends SettingBase {
    constructor(id, label, description, defaultTextValue, options, useUrlParams, defaultUrlParamResolver = function (value) {
        /* Return the string as-is by default */
        return value;
    }, defaultOnChangeListener = () => {
        /* Do nothing, to be overridden. */
    }) {
        super(id, label, description, defaultTextValue, defaultOnChangeListener);
        this._urlParamResolver = defaultUrlParamResolver;
        const stringToMatch = this.hasURLParam(this.id)
            ? this._urlParamResolver(this.getURLParam(this.id))
            : defaultTextValue;
        this.options = options !== null && options !== void 0 ? options : [stringToMatch];
        this.selected = stringToMatch;
        this.useUrlParams = useUrlParams;
    }
    getValueAsString() {
        return this.selected;
    }
    /**
     * Add a change listener to the select element.
     */
    addOnChangedListener(onChangedFunc) {
        this.onChange = onChangedFunc;
    }
    /**
     * @returns All available options as an array
     */
    get options() {
        return this._options;
    }
    /**
     * Set options
     * @param values Array of options
     */
    set options(values) {
        this._options = values;
        this.onChangeEmit(this.selected);
    }
    /**
     * @returns Selected option as a string
     */
    get selected() {
        return this.value;
    }
    /**
     * Set selected option if it matches one of the available options
     * @param value Selected option
     */
    set selected(value) {
        if (value === undefined) {
            return;
        }
        // If options contains the value, then set that as selected
        if (this.options.includes(value)) {
            this.value = value;
        }
        else {
            Logger.Error(`Could not set "${value}" as the selected option for ${this.id} because it wasn't one of the options.`);
        }
    }
    /**
     * Set the url parameter resolver to do some transformation to the string value
     * that is extracted from the url parameters.
     * @param urlParam A function that transforms the extracted url parameter string for this setting to something else.
     */
    set urlParamResolver(value) {
        this._urlParamResolver = value;
    }
}
