// @epicgames-ps/lib-pixelstreamingcommon-ue5.5/dist/cjs/pixelstreamingcommon.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeepaliveMonitor = exports.MessageHelpers = exports.Messages = exports.MessageRegistry = exports.EventEmitter = exports.SignallingProtocol = exports.WebSocketTransportNJS = exports.WebSocketTransport = exports.overrideLogger = exports.LogLevel = exports.Logger = void 0;
// Copyright Epic Games, Inc. All Rights Reserved.
var Logger_1 = require("./Logger/Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return Logger_1.Logger; } });
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return Logger_1.LogLevel; } });
Object.defineProperty(exports, "overrideLogger", { enumerable: true, get: function () { return Logger_1.overrideLogger; } });
var WebSocketTransport_1 = require("./Transport/WebSocketTransport");
Object.defineProperty(exports, "WebSocketTransport", { enumerable: true, get: function () { return WebSocketTransport_1.WebSocketTransport; } });
var WebSocketTransportNJS_1 = require("./Transport/WebSocketTransportNJS");
Object.defineProperty(exports, "WebSocketTransportNJS", { enumerable: true, get: function () { return WebSocketTransportNJS_1.WebSocketTransportNJS; } });
var SignallingProtocol_1 = require("./Protocol/SignallingProtocol");
Object.defineProperty(exports, "SignallingProtocol", { enumerable: true, get: function () { return SignallingProtocol_1.SignallingProtocol; } });
var EventEmitter_1 = require("./Event/EventEmitter");
Object.defineProperty(exports, "EventEmitter", { enumerable: true, get: function () { return EventEmitter_1.EventEmitter; } });
var message_registry_1 = require("./Messages/message_registry");
Object.defineProperty(exports, "MessageRegistry", { enumerable: true, get: function () { return message_registry_1.MessageRegistry; } });
exports.Messages = __importStar(require("./Messages/signalling_messages"));
exports.MessageHelpers = __importStar(require("./Messages/message_helpers"));
var KeepaliveMonitor_1 = require("./Protocol/KeepaliveMonitor");
Object.defineProperty(exports, "KeepaliveMonitor", { enumerable: true, get: function () { return KeepaliveMonitor_1.KeepaliveMonitor; } });
__exportStar(require("./Util/SdpUtils"), exports);
