// @epicgames-ps/lib-pixelstreamingfrontend-ue5.5/dist/cjs/pixelstreamingfrontend.js
"use strict";
// Copyright Epic Games, Inc. All Rights Reserved.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteOutboundRTPStats = exports.OutboundRTPStats = exports.InboundVideoStats = exports.InboundAudioStats = exports.DataChannelStats = exports.CandidateStat = exports.CandidatePairStats = exports.MessageDirection = exports.AggregatedStats = exports.LatencyTestResults = exports.AfkLogic = exports.PixelStreaming = exports.SettingText = exports.SettingOption = exports.SettingNumber = exports.SettingFlag = exports.SettingBase = exports.WebXRController = exports.WebRtcPlayerController = void 0;
var WebRtcPlayerController_1 = require("./WebRtcPlayer/WebRtcPlayerController");
Object.defineProperty(exports, "WebRtcPlayerController", { enumerable: true, get: function () { return WebRtcPlayerController_1.WebRtcPlayerController; } });
var WebXRController_1 = require("./WebXR/WebXRController");
Object.defineProperty(exports, "WebXRController", { enumerable: true, get: function () { return WebXRController_1.WebXRController; } });
__exportStar(require("./Config/Config"), exports);
var SettingBase_1 = require("./Config/SettingBase");
Object.defineProperty(exports, "SettingBase", { enumerable: true, get: function () { return SettingBase_1.SettingBase; } });
var SettingFlag_1 = require("./Config/SettingFlag");
Object.defineProperty(exports, "SettingFlag", { enumerable: true, get: function () { return SettingFlag_1.SettingFlag; } });
var SettingNumber_1 = require("./Config/SettingNumber");
Object.defineProperty(exports, "SettingNumber", { enumerable: true, get: function () { return SettingNumber_1.SettingNumber; } });
var SettingOption_1 = require("./Config/SettingOption");
Object.defineProperty(exports, "SettingOption", { enumerable: true, get: function () { return SettingOption_1.SettingOption; } });
var SettingText_1 = require("./Config/SettingText");
Object.defineProperty(exports, "SettingText", { enumerable: true, get: function () { return SettingText_1.SettingText; } });
var PixelStreaming_1 = require("./PixelStreaming/PixelStreaming");
Object.defineProperty(exports, "PixelStreaming", { enumerable: true, get: function () { return PixelStreaming_1.PixelStreaming; } });
var AFKController_1 = require("./AFK/AFKController");
Object.defineProperty(exports, "AfkLogic", { enumerable: true, get: function () { return AFKController_1.AFKController; } });
var LatencyTestResults_1 = require("./DataChannel/LatencyTestResults");
Object.defineProperty(exports, "LatencyTestResults", { enumerable: true, get: function () { return LatencyTestResults_1.LatencyTestResults; } });
__exportStar(require("./DataChannel/InitialSettings"), exports);
var AggregatedStats_1 = require("./PeerConnectionController/AggregatedStats");
Object.defineProperty(exports, "AggregatedStats", { enumerable: true, get: function () { return AggregatedStats_1.AggregatedStats; } });
__exportStar(require("./Util/InputCoordTranslator"), exports);
var StreamMessageController_1 = require("./UeInstanceMessage/StreamMessageController");
Object.defineProperty(exports, "MessageDirection", { enumerable: true, get: function () { return StreamMessageController_1.MessageDirection; } });
var CandidatePairStats_1 = require("./PeerConnectionController/CandidatePairStats");
Object.defineProperty(exports, "CandidatePairStats", { enumerable: true, get: function () { return CandidatePairStats_1.CandidatePairStats; } });
var CandidateStat_1 = require("./PeerConnectionController/CandidateStat");
Object.defineProperty(exports, "CandidateStat", { enumerable: true, get: function () { return CandidateStat_1.CandidateStat; } });
var DataChannelStats_1 = require("./PeerConnectionController/DataChannelStats");
Object.defineProperty(exports, "DataChannelStats", { enumerable: true, get: function () { return DataChannelStats_1.DataChannelStats; } });
var InboundRTPStats_1 = require("./PeerConnectionController/InboundRTPStats");
Object.defineProperty(exports, "InboundAudioStats", { enumerable: true, get: function () { return InboundRTPStats_1.InboundAudioStats; } });
Object.defineProperty(exports, "InboundVideoStats", { enumerable: true, get: function () { return InboundRTPStats_1.InboundVideoStats; } });
var OutBoundRTPStats_1 = require("./PeerConnectionController/OutBoundRTPStats");
Object.defineProperty(exports, "OutboundRTPStats", { enumerable: true, get: function () { return OutBoundRTPStats_1.OutboundRTPStats; } });
Object.defineProperty(exports, "RemoteOutboundRTPStats", { enumerable: true, get: function () { return OutBoundRTPStats_1.RemoteOutboundRTPStats; } });
__exportStar(require("./PeerConnectionController/LatencyCalculator"), exports);
__exportStar(require("./DataChannel/DataChannelLatencyTestResults"), exports);
__exportStar(require("./Util/EventEmitter"), exports);
__exportStar(require("@epicgames-ps/lib-pixelstreamingcommon-ue5.5"), exports);
