// @epicgames-ps/lib-pixelstreamingfrontend-ui-ue5.5/dist/cjs/Config/ConfigUI.js
"use strict";
// Copyright Epic Games, Inc. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigUI = void 0;
const lib_pixelstreamingfrontend_ue5_5_1 = require("@epicgames-ps/lib-pixelstreamingfrontend-ue5.5");
const SettingUIFlag_1 = require("./SettingUIFlag");
const SettingUINumber_1 = require("./SettingUINumber");
const SettingUIText_1 = require("./SettingUIText");
const SettingUIOption_1 = require("./SettingUIOption");
const UIConfigurationTypes_1 = require("../UI/UIConfigurationTypes");
class ConfigUI {
    // ------------ Settings -----------------
    constructor(config) {
        this.customFlags = new Map();
        /* A map of flags that can be toggled - options that can be set in the application - e.g. Use Mic? */
        this.flagsUi = new Map();
        /* A map of numerical settings - options that can be in the application - e.g. MinBitrate */
        this.numericParametersUi = new Map();
        /* A map of text settings - e.g. signalling server url */
        this.textParametersUi = new Map();
        /* A map of enum based settings - e.g. preferred codec */
        this.optionParametersUi = new Map();
        this.createCustomUISettings(config.useUrlParams);
        this.registerSettingsUIComponents(config);
    }
    /**
     * Create custom UI settings that are not provided by the Pixel Streaming library.
     */
    createCustomUISettings(useUrlParams) {
        this.customFlags.set(UIConfigurationTypes_1.ExtraFlags.LightMode, new lib_pixelstreamingfrontend_ue5_5_1.SettingFlag(UIConfigurationTypes_1.ExtraFlags.LightMode, 'Color Scheme: Dark Mode', 'Page styling will be either light or dark', false /*if want to use system pref: (window.matchMedia && window.matchMedia('(prefers-color-scheme: light)').matches)*/, useUrlParams, (isLightMode, setting) => {
            setting.label = `Color Scheme: ${isLightMode ? 'Light' : 'Dark'} Mode`;
        }));
    }
    /**
     * Creates UI wrapper components for each setting element in config.
     * @param config -
     */
    registerSettingsUIComponents(config) {
        for (const setting of config.getFlags()) {
            this.flagsUi.set(setting.id, new SettingUIFlag_1.SettingUIFlag(setting));
        }
        for (const setting of Array.from(this.customFlags.values())) {
            this.flagsUi.set(setting.id, new SettingUIFlag_1.SettingUIFlag(setting));
        }
        for (const setting of config.getTextSettings()) {
            this.textParametersUi.set(setting.id, new SettingUIText_1.SettingUIText(setting));
        }
        for (const setting of config.getNumericSettings()) {
            this.numericParametersUi.set(setting.id, new SettingUINumber_1.SettingUINumber(setting));
        }
        for (const setting of config.getOptionSettings()) {
            this.optionParametersUi.set(setting.id, new SettingUIOption_1.SettingUIOption(setting));
        }
    }
    /**
     * Make DOM elements for a settings section with a heading.
     * @param settingsElem - The parent container for our DOM elements.
     * @param sectionHeading - The heading element to go into the section.
     * @returns The constructed DOM element for the section.
     */
    buildSectionWithHeading(settingsElem, sectionHeading) {
        // make section element
        const sectionElem = document.createElement('section');
        sectionElem.classList.add('settingsContainer');
        // make section heading
        const psSettingsHeader = document.createElement('div');
        psSettingsHeader.classList.add('settingsHeader');
        psSettingsHeader.classList.add('settings-text');
        psSettingsHeader.textContent = sectionHeading;
        // add section and heading to parent settings element
        sectionElem.appendChild(psSettingsHeader);
        settingsElem.appendChild(sectionElem);
        return sectionElem;
    }
    /**
     * Setup flags with their default values and add them to the `Config.flags` map.
     * @param settingsElem - - The element that contains all the individual settings sections, flags, and so on.
     */
    populateSettingsElement(settingsElem, settingsConfig) {
        if ((0, UIConfigurationTypes_1.isSectionEnabled)(settingsConfig, UIConfigurationTypes_1.SettingsSections.PixelStreaming)) {
            /* Setup all Pixel Streaming specific settings */
            const psSettingsSection = this.buildSectionWithHeading(settingsElem, UIConfigurationTypes_1.SettingsSections.PixelStreaming);
            // make settings show up in DOM
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.TextParameters.SignallingServerUrl))
                this.addSettingText(psSettingsSection, this.textParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.TextParameters.SignallingServerUrl));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.OptionParameters.StreamerId))
                this.addSettingOption(psSettingsSection, this.optionParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.OptionParameters.StreamerId));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.AutoConnect))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.AutoConnect));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.AutoPlayVideo))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.AutoPlayVideo));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.UseMic))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.UseMic));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.UseCamera))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.UseCamera));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.StartVideoMuted))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.StartVideoMuted));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.IsQualityController))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.IsQualityController));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.ForceMonoAudio))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.ForceMonoAudio));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.ForceTURN))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.ForceTURN));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.SuppressBrowserKeys))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.SuppressBrowserKeys));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.AFKDetection))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.AFKDetection));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.WaitForStreamer))
                this.addSettingFlag(psSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.WaitForStreamer));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.AFKTimeoutSecs))
                this.addSettingNumeric(psSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.AFKTimeoutSecs));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.AFKCountdownSecs))
                this.addSettingNumeric(psSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.AFKCountdownSecs));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.MaxReconnectAttempts))
                this.addSettingNumeric(psSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.MaxReconnectAttempts));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.StreamerAutoJoinInterval))
                this.addSettingNumeric(psSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.StreamerAutoJoinInterval));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.KeepaliveDelay))
                this.addSettingNumeric(psSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.KeepaliveDelay));
        }
        if ((0, UIConfigurationTypes_1.isSectionEnabled)(settingsConfig, UIConfigurationTypes_1.SettingsSections.UI)) {
            /* Setup all view/ui related settings under this section */
            const viewSettingsSection = this.buildSectionWithHeading(settingsElem, UIConfigurationTypes_1.SettingsSections.UI);
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.MatchViewportResolution))
                this.addSettingFlag(viewSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.MatchViewportResolution));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.HoveringMouseMode))
                this.addSettingFlag(viewSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.HoveringMouseMode));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, UIConfigurationTypes_1.ExtraFlags.LightMode))
                this.addSettingFlag(viewSettingsSection, this.flagsUi.get(UIConfigurationTypes_1.ExtraFlags.LightMode));
        }
        if ((0, UIConfigurationTypes_1.isSectionEnabled)(settingsConfig, UIConfigurationTypes_1.SettingsSections.Input)) {
            /* Setup all encoder related settings under this section */
            const inputSettingsSection = this.buildSectionWithHeading(settingsElem, UIConfigurationTypes_1.SettingsSections.Input);
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.KeyboardInput))
                this.addSettingFlag(inputSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.KeyboardInput));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.MouseInput))
                this.addSettingFlag(inputSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.MouseInput));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.FakeMouseWithTouches))
                this.addSettingFlag(inputSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.FakeMouseWithTouches));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.TouchInput))
                this.addSettingFlag(inputSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.TouchInput));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.GamepadInput))
                this.addSettingFlag(inputSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.GamepadInput));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.Flags.XRControllerInput))
                this.addSettingFlag(inputSettingsSection, this.flagsUi.get(lib_pixelstreamingfrontend_ue5_5_1.Flags.XRControllerInput));
        }
        if ((0, UIConfigurationTypes_1.isSectionEnabled)(settingsConfig, UIConfigurationTypes_1.SettingsSections.Encoder)) {
            /* Setup all encoder related settings under this section */
            const encoderSettingsSection = this.buildSectionWithHeading(settingsElem, UIConfigurationTypes_1.SettingsSections.Encoder);
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.CompatQualityMin))
                this.addSettingNumeric(encoderSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.CompatQualityMin));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.CompatQualityMax))
                this.addSettingNumeric(encoderSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.CompatQualityMax));
            const preferredCodecOption = this.optionParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.OptionParameters.PreferredCodec);
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.OptionParameters.PreferredCodec))
                this.addSettingOption(encoderSettingsSection, this.optionParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.OptionParameters.PreferredCodec));
            if (preferredCodecOption &&
                [...preferredCodecOption.selector.options]
                    .map((o) => o.value)
                    .includes('Only available on Chrome')) {
                preferredCodecOption.disable();
            }
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.OptionParameters.PreferredQuality))
                this.addSettingOption(encoderSettingsSection, this.optionParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.OptionParameters.PreferredQuality));
        }
        if ((0, UIConfigurationTypes_1.isSectionEnabled)(settingsConfig, UIConfigurationTypes_1.SettingsSections.WebRTC)) {
            /* Setup all webrtc related settings under this section */
            const webrtcSettingsSection = this.buildSectionWithHeading(settingsElem, UIConfigurationTypes_1.SettingsSections.WebRTC);
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.WebRTCFPS))
                this.addSettingNumeric(webrtcSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.WebRTCFPS));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.WebRTCMinBitrate))
                this.addSettingNumeric(webrtcSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.WebRTCMinBitrate));
            if ((0, UIConfigurationTypes_1.isSettingEnabled)(settingsConfig, lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.WebRTCMaxBitrate))
                this.addSettingNumeric(webrtcSettingsSection, this.numericParametersUi.get(lib_pixelstreamingfrontend_ue5_5_1.NumericParameters.WebRTCMaxBitrate));
        }
    }
    /**
     * Add a SettingText element to a particular settings section in the DOM and registers that text in the text settings map.
     * @param settingsSection - The settings section HTML element.
     * @param settingText - The textual settings object.
     */
    addSettingText(settingsSection, settingText) {
        if (settingText) {
            settingsSection.appendChild(settingText.rootElement);
            this.textParametersUi.set(settingText.setting.id, settingText);
        }
    }
    /**
     * Add a SettingFlag element to a particular settings section in the DOM and registers that flag in the Config.flag map.
     * @param settingsSection - The settings section HTML element.
     * @param settingFlag - The settings flag object.
     */
    addSettingFlag(settingsSection, settingFlag) {
        if (settingFlag) {
            settingsSection.appendChild(settingFlag.rootElement);
            this.flagsUi.set(settingFlag.setting.id, settingFlag);
        }
    }
    /**
     * Add a numeric setting element to a particular settings section in the DOM and registers that flag in the Config.numericParameters map.
     * @param settingsSection - The settings section HTML element.
     * @param settingFlag - The settings flag object.
     */
    addSettingNumeric(settingsSection, setting) {
        if (setting) {
            settingsSection.appendChild(setting.rootElement);
            this.numericParametersUi.set(setting.setting.id, setting);
        }
    }
    /**
     * Add an enum based settings element to a particular settings section in the DOM and registers that flag in the Config.enumParameters map.
     * @param settingsSection - The settings section HTML element.
     * @param settingFlag - The settings flag object.
     */
    addSettingOption(settingsSection, setting) {
        if (setting) {
            settingsSection.appendChild(setting.rootElement);
            this.optionParametersUi.set(setting.setting.id, setting);
        }
    }
    onSettingsChanged({ data: { id, target, type } }) {
        if (type === 'flag') {
            const _id = id;
            const _target = target;
            const setting = this.flagsUi.get(_id);
            if (setting) {
                if (setting.flag !== _target.flag) {
                    setting.flag = _target.flag;
                }
                if (setting.label !== _target.label) {
                    setting.label = _target.label;
                }
            }
        }
        else if (type === 'number') {
            const _id = id;
            const _target = target;
            const setting = this.numericParametersUi.get(_id);
            if (setting) {
                if (setting.number !== _target.number) {
                    setting.number = _target.number;
                }
                if (setting.label !== _target.label) {
                    setting.label = _target.label;
                }
            }
        }
        else if (type === 'text') {
            const _id = id;
            const _target = target;
            const setting = this.textParametersUi.get(_id);
            if (setting) {
                if (setting.text !== _target.text) {
                    setting.text = _target.text;
                }
                if (setting.label !== _target.label) {
                    setting.label = _target.label;
                }
            }
        }
        else if (type === 'option') {
            const _id = id;
            const _target = target;
            const setting = this.optionParametersUi.get(_id);
            if (setting) {
                const uiOptions = setting.options;
                const targetOptions = _target.options;
                if (uiOptions.length !== targetOptions.length ||
                    !uiOptions.every((value) => targetOptions.includes(value))) {
                    setting.options = _target.options;
                }
                if (setting.selected !== _target.selected) {
                    setting.selected = _target.selected;
                }
                if (setting.label !== _target.label) {
                    setting.label = _target.label;
                }
            }
        }
    }
    /**
     * Add a callback to fire when the flag is toggled.
     * @param id - The id of the flag.
     * @param onChangeListener - The callback to fire when the value changes.
     */
    addCustomFlagOnSettingChangedListener(id, onChangeListener) {
        if (this.customFlags.has(id)) {
            this.customFlags.get(id).onChange = onChangeListener;
        }
    }
    /**
     * Set the label for the flag.
     * @param id - The id of the flag.
     * @param label - The new label to use for the flag.
     */
    setCustomFlagLabel(id, label) {
        if (!this.customFlags.has(id)) {
            lib_pixelstreamingfrontend_ue5_5_1.Logger.Warning(`Cannot set label for flag called ${id} - it does not exist in the Config.flags map.`);
        }
        else {
            this.customFlags.get(id).label = label;
            this.flagsUi.get(id).label = label;
        }
    }
    /**
     * Get the value of the configuration flag which has the given id.
     * @param id - The unique id for the flag.
     * @returns True if the flag is enabled.
     */
    isCustomFlagEnabled(id) {
        return this.customFlags.get(id).flag;
    }
    disableSetting(id) {
        var _a, _b, _c, _d;
        if ((0, lib_pixelstreamingfrontend_ue5_5_1.isFlagId)(id)) {
            (_a = this.flagsUi.get(id)) === null || _a === void 0 ? void 0 : _a.disable();
        }
        else if ((0, lib_pixelstreamingfrontend_ue5_5_1.isNumericId)(id)) {
            (_b = this.numericParametersUi.get(id)) === null || _b === void 0 ? void 0 : _b.disable();
        }
        else if ((0, lib_pixelstreamingfrontend_ue5_5_1.isTextId)(id)) {
            (_c = this.textParametersUi.get(id)) === null || _c === void 0 ? void 0 : _c.disable();
        }
        else if ((0, lib_pixelstreamingfrontend_ue5_5_1.isOptionId)(id)) {
            (_d = this.optionParametersUi.get(id)) === null || _d === void 0 ? void 0 : _d.disable();
        }
    }
    enableSetting(id) {
        var _a, _b, _c, _d;
        if ((0, lib_pixelstreamingfrontend_ue5_5_1.isFlagId)(id)) {
            (_a = this.flagsUi.get(id)) === null || _a === void 0 ? void 0 : _a.enable();
        }
        else if ((0, lib_pixelstreamingfrontend_ue5_5_1.isNumericId)(id)) {
            (_b = this.numericParametersUi.get(id)) === null || _b === void 0 ? void 0 : _b.enable();
        }
        else if ((0, lib_pixelstreamingfrontend_ue5_5_1.isTextId)(id)) {
            (_c = this.textParametersUi.get(id)) === null || _c === void 0 ? void 0 : _c.enable();
        }
        else if ((0, lib_pixelstreamingfrontend_ue5_5_1.isOptionId)(id)) {
            (_d = this.optionParametersUi.get(id)) === null || _d === void 0 ? void 0 : _d.enable();
        }
    }
}
exports.ConfigUI = ConfigUI;
