// @epicgames-ps/lib-pixelstreamingcommon-ue5.5/build/commonjs/pixelstreamingcommon.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageHelpers = exports.Messages = exports.MessageRegistry = exports.SignallingProtocol = exports.WebSocketTransportNJS = exports.WebSocketTransport = exports.LogLevel = exports.Logger = void 0;
var Logger_1 = require("./Logger/Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return Logger_1.Logger; } });
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return Logger_1.LogLevel; } });
var WebSocketTransport_1 = require("./Transport/WebSocketTransport");
Object.defineProperty(exports, "WebSocketTransport", { enumerable: true, get: function () { return WebSocketTransport_1.WebSocketTransport; } });
var WebSocketTransportNJS_1 = require("./Transport/WebSocketTransportNJS");
Object.defineProperty(exports, "WebSocketTransportNJS", { enumerable: true, get: function () { return WebSocketTransportNJS_1.WebSocketTransportNJS; } });
var SignallingProtocol_1 = require("./Protocol/SignallingProtocol");
Object.defineProperty(exports, "SignallingProtocol", { enumerable: true, get: function () { return SignallingProtocol_1.SignallingProtocol; } });
var message_registry_1 = require("./Messages/message_registry");
Object.defineProperty(exports, "MessageRegistry", { enumerable: true, get: function () { return message_registry_1.MessageRegistry; } });
exports.Messages = __importStar(require("./Messages/signalling_messages"));
exports.MessageHelpers = __importStar(require("./Messages/message_helpers"));
