// @epicgames-ps/lib-pixelstreamingfrontend-ui-ue5.5/dist/esm/UI/Controls.js
// Copyright Epic Games, Inc. All Rights Reserved.
import { FullScreenIcon } from './FullscreenIcon';
import { SettingsIcon } from './SettingsIcon';
import { StatsIcon } from './StatsIcon';
import { XRIcon } from './XRIcon';
import { WebXRController } from '@epicgames-ps/lib-pixelstreamingfrontend-ue5.5';
import { UIElementCreationMode } from '../UI/UIConfigurationTypes';
// If there isn't a type provided, default behaviour is to create the element.
function shouldCreateButton(type) {
    return type == undefined ? true : type.creationMode === UIElementCreationMode.CreateDefaultElement;
}
/**
 * Element containing various controls like stats, settings, fullscreen.
 */
export class Controls {
    /**
     * Construct the controls
     */
    constructor(config) {
        this.config = config;
        if (!config || shouldCreateButton(config.statsButtonType)) {
            this.statsIcon = new StatsIcon();
        }
        if (!config || shouldCreateButton(config.settingsButtonType)) {
            this.settingsIcon = new SettingsIcon();
        }
        if (!config || shouldCreateButton(config.fullscreenButtonType)) {
            this.fullscreenIcon = new FullScreenIcon();
        }
        if (!config || shouldCreateButton(config.xrIconType)) {
            this.xrIcon = new XRIcon();
        }
    }
    /**
     * Get the element containing the controls.
     */
    get rootElement() {
        if (!this._rootElement) {
            this._rootElement = document.createElement('div');
            this._rootElement.id = 'controls';
            if (this.fullscreenIcon) {
                this._rootElement.appendChild(this.fullscreenIcon.rootElement);
            }
            if (this.settingsIcon) {
                this._rootElement.appendChild(this.settingsIcon.rootElement);
            }
            if (this.statsIcon) {
                this._rootElement.appendChild(this.statsIcon.rootElement);
            }
            if (this.xrIcon) {
                void WebXRController.isSessionSupported('immersive-vr').then((supported) => {
                    if (supported) {
                        this._rootElement.appendChild(this.xrIcon.rootElement);
                    }
                });
            }
            document.addEventListener('fullscreenchange', this.handleFullscreenChange.bind(this));
        }
        return this._rootElement;
    }
    handleFullscreenChange() {
        var _a;
        const isInFullscreen = !!document.fullscreenElement;
        if (isInFullscreen && ((_a = this.config) === null || _a === void 0 ? void 0 : _a.hideControlsInFullscreen)) {
            this._rootElement.style.visibility = 'hidden';
        }
        else {
            this._rootElement.style.visibility = 'visible';
        }
    }
}
